### PR TITLE
Fixing mlas unittest failures in POWER

### PR DESCRIPTION
**Pull request by austinpagan**
We noted that "onnx_runtime_mlas_test" was reporting errors on our Power servers.

We discovered some places where error tolerances were too tight.
In one case, the demand was an EXACT MATCH.
In another, the tolerance was 0.000001 on ABSOLUTE ERROR, even though the numbers being compared were > 10^5.

We also discovered problems in how NaNs were being processed, and we have created more robust code for that.

We hope you find these changes acceptable.

### CODE REVIEW

File: onnxruntime/core/mlas/lib/tanh.cpp [hunk 1]
```diff
@@ -119,6 +119,11 @@ Return Value:
 
         float Value = *Input++;
 
+	// This odd "if test" exists to ensure an input value of NaN carries through
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/tanh.cpp:122]: >	 [](http://example.com/codeflow?start=0&length=1)

nit: space

File: onnxruntime/core/mlas/lib/logistic.cpp [hunk 2]
```diff
@@ -121,6 +121,11 @@ Return Value:
 
         float Value = *Input++;
 
+	// This odd "if test" exists to ensure an input value of NaN carries through
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/logistic.cpp:124]: >	 [](http://example.com/codeflow?start=0&length=1)

nit:space

File: onnxruntime/core/mlas/lib/logistic.cpp [hunk 3]
```diff
@@ -121,6 +121,11 @@ Return Value:
...
+        // without modification because "std::min" and "std::max" return unreliable results
+        // when NaNs are involved, and it's clear from the test's reference outputs that
+        // they want a NaN on output whenever the input is a NaN.
+        if (Value == Value)
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/logistic.cpp:128]: >if (Value == Value) [](http://example.com/codeflow?start=8&length=19)

this is odd. NaN is sticky. NaN is the expected result for any operatation on NaN, right?

Comment (austinpagan) [reply onnxruntime/core/mlas/lib/logistic.cpp:128]: Yes, you'd like to think so, but std::min and std::max turn out to be exceptions.

They end up doing a **_comparison_** on the two operands.  
As it happens, if a NaN is involved in either A or B, then (A < B) and (A > B) always (**_ALWAYS_**) return FALSE.  
This makes std::min and std::max return really unfortunate results.
For example, std::min(3.0, NaN) will return NaN, but std::min(NaN, 3.0) will return 3.0.

Basically, one should NOT use std::min or std::max if an input value _might_ be NaN.

Here's a web page which explores this in more depth: 
https://microsoft.public.vc.stl.narkive.com/KmbP2DlQ/std-max-and-nan

So, for both tanh.cpp and logistic.cpp, I just put in that "(Value == Value)" thing, so if Value is a NaN, then the output Value stays NaN (which is the desired result).

File: onnxruntime/core/mlas/lib/tanh.cpp (hunk 1 above)
Comment (austinpagan) [reply onnxruntime/core/mlas/lib/tanh.cpp:122]: I'm sorry, I get you are saying "nit" as in "this is a small complaint" but when you say "space," I do not understand what you are requesting...

File: onnxruntime/core/mlas/lib/logistic.cpp (hunk 2 above)
Comment (austinpagan) [reply onnxruntime/core/mlas/lib/logistic.cpp:124]: (see above)

File: onnxruntime/core/mlas/lib/tanh.cpp (hunk 1 above)
Comment (austinpagan) [reply onnxruntime/core/mlas/lib/tanh.cpp:122]: This line used a tab and the others used spaces.  When I edit it, it looks fine.
I will change it to use spaces as well.

File: onnxruntime/core/mlas/lib/logistic.cpp (hunk 2 above)
Comment (austinpagan) [reply onnxruntime/core/mlas/lib/logistic.cpp:124]: This line used a tab and the others used spaces.  When I edit it, it looks fine.
I will change it to use spaces as well.

File: onnxruntime/core/mlas/lib/logistic.cpp (hunk 3 above)
Comment (yufenglee) [reply onnxruntime/core/mlas/lib/logistic.cpp:128]: Thanks for the info. Then I think it will work to adjust the order of the statement below like this:
Value = std::min(std::max(Value, MlasLogisticConstants.LowerRange), MlasLogisticConstants.UpperRange);

Comment (austinpagan) [reply onnxruntime/core/mlas/lib/logistic.cpp:128]: Thank you for this suggestion.  I will give it more thought tomorrow when my brain is fresh, and maybe code that in and compile it and try it.  Stay tuned!

Comment (yufenglee) [reply onnxruntime/core/mlas/lib/logistic.cpp:128]: Based on this doc: https://en.cppreference.com/w/cpp/algorithm/max . std::max is implemented as `(a < b) ? b : a; ` , so if a is nan, then it returns nan. similar to min.

We can also replace std::max with `(a < b) ? b : a; ` explicitly, in case std::max is not implemented consistently across different library.

Comment (austinpagan) [reply onnxruntime/core/mlas/lib/logistic.cpp:128]: I'm always a fan of using simpler calls rather than library calls.
So, I guess we could do this in two steps:
```
   float v_tmp;
   v_tmp = (Value < MlasLogisticConstants.LowerRange) ? MlasLogisticConstants.LowerRange : Value;
   Value = (v_tmp > MlasLogisticConstants.UpperRange) ? MlasLogisticConstants.UpperRange : v_tmp;
```

**Comment by austinpagan** [comment]
Just FYI, I'm hoping for a reply from @yufenglee on my proposal from 3 days ago (see above)...

**Comment by snnn** [comment]
I just pinged him. Sorry, he is out of town now.  He will review your change when he is back.

**Review by yufenglee — APPROVED** [review]
